### PR TITLE
Thermal model

### PR DESCRIPTION
**Pull request by Siddharth-Barve**
Incorporated Thermal Model

Thermal Model:

- Grid cells interact with adjacent cells
- Agent core (center) cells generate energy (Metabolism)
- Air cells undergo environmental cooling (Newton's Law of Cooling)
- Heat exchange is modeled by evaluating total thermal resistivity and temperature difference between cells
- Temperature change is proportional to heat exchanged and inversely proportional to mass and specific heat capacity

Agents:

- Changed body temp of core (center) to numpy array of each agent grid cell temp (alive/dead still uses core)
- Specific heat of agents is currently constant and not a parameter
- Height of agents is currently constant and not a parameter
- Outer most cells of agents are considered internal but involve insulation (external heat conductivity) when interacting with grid cells not belonging to the agent

Environment -> update_thermal:

- Local agent id numpy array included
- Local heat capacity map included
- Local previous material map included to fill gaps left by moving penguins with air

### CODE REVIEW

**Comment by Siddharth-Barve** [comment]
Edit:

Environment:

- Specific heat and density of air is currently constant and not a parameter

**Review by liu2g — APPROVED** [review]

File: src/agent.py [hunk 1]
```diff
@@ -50,7 +50,7 @@ def __init__(
         self._col = col
         self._body_radius = body_radius
         self._sense_radius = sense_radius
-        self._body_temp = body_temp
+        self._body_temp = np.full(shape=(2*body_radius-1,2*body_radius-1), fill_value=body_temp, dtype=float)
```
Comment (stegnerw) [inline src/agent.py:53]: Looks like this produces a square of heat instead of a diamond shape. Is that okay?

File: src/environment.py [hunk 2]
```diff
@@ -77,10 +77,17 @@ def __init__(
         )
 
         # Thermal model related members
-        self._thermal_map = np.empty(shape=self._env_size, dtype=float)
+        self._thermal_map = np.full(shape=self._env_size, fill_value=initial_air_temp, dtype=float)
         # I intend this to store a string but Sid you may change the dtype
```
Comment (stegnerw) [inline src/environment.py:81]: This comment should be removed.

File: src/environment.py [hunk 3]
```diff
@@ -77,10 +77,17 @@ def __init__(
...
         # I did not do enums because I dislike Python enums
```
Comment (stegnerw) [inline src/environment.py:82]: This comment should be removed.

File: src/environment.py [hunk 4]
```diff
@@ -77,10 +77,17 @@ def __init__(
...
-        self._material_map = np.empty(shape=self._env_size, dtype=object)
+        """
+        Material Map Key:
+            [0] = Air
+            [1] = Penguin Core
+            [2] = Penguin Internal
+            [3] = Penguin External
+        """
+        self._material_map = np.zeros(shape=self._env_size, dtype=float)
```
Comment (stegnerw) [inline src/environment.py:90]: Is there a reason this is `float` instead of `int`?

File: src/environment.py [hunk 5]
```diff
@@ -112,15 +119,163 @@ def run(self) -> None:
...
 
         This will include calculating the temperature of every tile,
         calculating the body temperature of each agent, and anything
         else included in the thermal model.
         """
-        # TODO: Implement this
+        
+        prev_material_map=self._material_map
+        
+        """Update Maps"""
```
Comment (stegnerw) [inline src/environment.py:132]: I'm going to assume this is correct and we can go over it later.

**Review by stegnerw — APPROVED** [review]
My comments were minor so we can just merge it and clean up comments and stuff later.

**Pull request merged**